R/join_keys.R

```diff
@@ -0,0 +1,651 @@
+    return(value)
+  }
+
+  if (inherits(value, "join_key_set")) value <- list(value)
+
+  checkmate::assert_list(value, types = "join_key_set", min.len = 1)
+
+  join_keys_obj <- new_join_keys()
+
+  # check if any join_key_sets share the same datasets but different values
+  for (idx_1 in seq_along(value)) {
+    for (idx_2 in seq_along(value[idx_1])) {
+      assert_compatible_keys(value[[idx_1]], value[[idx_2]])
+    }
+    dataset_1 <- get_dataset_1(value[[idx_1]])
+    dataset_2 <- get_dataset_2(value[[idx_1]])
+    keys <- get_keys(value[[idx_1]])
+
+    join_keys_obj[[dataset_1]][[dataset_2]] <- keys
+  }
+
+  logger::log_trace("join_keys keys are set.")
+
+  join_keys_obj
+}
+
+#' @rdname join_keys
+#' @export
+#'
+#' @examples
+#'
+#' c(join_keys(join_key("a", "b", "c")), join_keys(join_key("a", "d2", "c")))
+c.join_keys <- function(...) {
+  x <- rlang::list2(...)
+
+  if (!length(x)) {
+    return(NULL)
+  }
+  checkmate::assert_list(x[-1], types = c("join_keys", "join_key_set"))
+
+  merge_join_keys.join_keys(x[[1]], x[-1])
+}
+
+#' @rdname join_keys
+#' @export
+#' @examples
+#'
+#' # Setter for join_keys within teal_data ----
+#'
+#' td <- teal_data()
+#' join_keys(td)[["ds1"]][["ds2"]] <- "key1"
+#' join_keys(td)[["ds2"]][["ds2"]] <- "key2"
+#' join_keys(td) <- c(join_keys(td), join_keys(join_key("ds3", "ds2", "key3")))
+#' join_keys(td)
+`join_keys<-.teal_data` <- function(join_keys_obj, value) {
+  if (missing(value)) {
+    return(join_keys_obj)
+  }
+
+  if (checkmate::test_class(value, c("join_keys", "list"))) {
+    join_keys_obj@join_keys <- merge_join_keys(join_keys_obj@join_keys, value)
+    return(join_keys_obj)
+  }
+
+  join_keys(join_keys_obj@join_keys) <- value
+  join_keys_obj
+}
+
+#' The Names of an `join_keys` Object
+#' @inheritParams base::`names<-`
+#' @export
+`names<-.join_keys` <- function(x, value) {
+  x <- unclass(x)
+  # Update inner keys
+  for (old_name in setdiff(names(x), value)) {
+    old_entry <- x[[old_name]]
+    new_name <- value[names(x) == old_name]
+
+    # Change 2nd-tier first
+    for (sub_name in names(old_entry)) {
+      names(x[[sub_name]])[names(x[[sub_name]]) == old_name] <- new_name
+    }
+
+    # Change in first tier
+    names(x)[names(x) == old_name] <- new_name
+  }
+  class(x) <- c("join_keys", "list")
+  x
+}
+
+#' @rdname join_keys
+#' @details
+#' Getter for `join_keys` that returns the relationship between pairs of datasets.
+#'
+#' @param join_keys_obj (`join_keys`) object to extract the join keys
+#' @param dataset_1 (`character`) name of first dataset.
+#'
+#' @export
+#'
+#' @examples
+#'
+#' # Getter for join_keys ----
+#'
+#' jk <- join_keys()
+#' jk[["ds1"]][["ds2"]] <- "some_col"
+#' jk[["ds1"]][["ds3"]] <- "some_col2"
+#'
+#' jk["ds1"]
+#' jk[1:2]
+#' jk[c("ds1", "ds2")]
+`[.join_keys` <- function(join_keys_obj, dataset_1 = NULL) {
+  # Protection against missing being passed through functions
+  if (missing(dataset_1)) dataset_1 <- NULL
```

Comment:
If it can be `missing`  from other calls, why allow `NULL` and not just handle `missing`?

Reply:
Formals are identical to `[` so no more NULL (other than a call with `dataset_1` = NULL - now `i = NULL` - that returns an empty

``` r
aa <- list(1, 2, 3)
aa[NULL]
#> list()
jk <- join_keys(join_key("a", "b", "c"))
jk[NULL]
#> An empty join_keys object.
```